storage: efficient abort cache GC

tbg opened the pull request:
Implement a TODO: delete sequence cache entries solely based on the
last client activity on the Txn at abort time.
To be 100% correct, this requires a corresponding update to the coordinator to
prevent client activity after a failed heartbeat (see second commit):

To effectively garbage-collect abort cache entries, it is important to know
that a client who is running a transaction which has been aborted finds out
about this in a timely manner. This mechanism relies on the last heartbeat
timestamp observed at the time of the abort.

This chance also brings the coordinator side on par with this assumption: When
a heartbeat comes back with a non-PENDING transaction (or fails, in which case
we assume ABORTED), we lock down the coordinator for that client and serve it
a `TransactionAbortedError` directly from the coordinator. This error contains
the latest information from the transaction record, so it plays along nicely
with priority upgrades and the like.
In the process, fix an issue with mutating ba.Txn in `(*DistSender).Send`.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5882)

<!-- Reviewable:end -->

tbg committed:
storage: efficient abort cache GC

Implement a TODO: delete sequence cache entries solely based on the
last client activity on the Txn at abort time.
To be 100% correct, this requires a corresponding update to the coordinator to
prevent client activity after a failed heartbeat which is left as a follow-up
change.

tbg committed:
kv: prevent client activity based on heartbeat

To effectively garbage-collect abort cache entries, it is important to know
that a client who is running a transaction which has been aborted finds out
about this in a timely manner. This mechanism relies on the last heartbeat
timestamp observed at the time of the abort.

This chance brings the coordinator side on par with this assumption: When
a heartbeat comes back with a non-PENDING transaction (or fails, in which case
we assume ABORTED), we lock down the coordinator for that client and serve it
a `TransactionAbortedError` directly from the coordinator. This error contains
the latest information from the transaction record, so it plays along nicely
with priority upgrades and the like.

tbg committed:
kv: avoid mutating ba.Txn

This is not the cleanest fix since it does unnecessary work in some cases.
However, it removes an immediate test oddity and is best revisited in a
follow-up (along with corresponding code paths in `(*Replica)`).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=5882) <br/>You should sign our Contributor License Agreement in order to get your pull request merged.<br/>

tbg committed:
kv: Do not share Txn proto with client

TxnCoordSender was sharing a transaction proto from its
state with the client, which led to data races in the
previous commit.

tamird commented:
Reviewed 7 of 7 files at r1, 3 of 3 files at r2, 3 of 3 files at r3, 3 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[kv/dist_sender.go, line 488 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfeOy_jLz0OCWFoTYT:-KEfeOyajPLd2gc7fOaF:-938122477) ([raw file](https://github.com/cockroachdb/cockroach/blob/1888be7ba43df1771ccd175da270518f3ccd2a2e/kv/dist_sender.go#L488)):*
nit: you can combine this with the following `if` clause and nest the `len(...) == 0` block

---

*[kv/dist_sender.go, line 573 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEffMOzDaz8eLckb067:-KEffMOzDaz8eLckb068:-2137643989) ([raw file](https://github.com/cockroachdb/cockroach/blob/3220ac5b9a40c0ce6a3f9645c10978f055fe63d7/kv/dist_sender.go#L573)):*
this is the last remaining reference to Header() - maybe kill it?

---

*[kv/txn_coord_sender.go, line 643 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfd1U16RjKSeAFHBIS:-KEfd1U16RjKSeAFHBIT:-1766073256) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L643)):*
remove and use `:=` below?

---

*[kv/txn_coord_sender.go, line 660 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdCSRfZ9gorckxyCE:-KEfdCSSQ2XGO_RxLvoV:-1484360235) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L660)):*
empty block

---

*[kv/txn_coord_sender.go, line 731 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdpUazETUX7lWtKjz:-KEfdpUazETUX7lWtKk-:2101548610) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L731)):*
this is just `else`

---

*[kv/txn_coord_sender.go, line 742 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdjm7b4pwfxL9WuCN:-KEfdjm7b4pwfxL9WuCO:2095615263) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L742)):*
s/tem/them/

---

*[kv/txn_coord_sender.go, line 736 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfelkfXZt_AwrBlwyR:-KEfelkfXZt_AwrBlwyS:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/3220ac5b9a40c0ce6a3f9645c10978f055fe63d7/kv/txn_coord_sender.go#L736)):*
?

---

*[kv/txn_coord_sender_test.go, line 340 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdydf7AF77XII7HZf:-KEfdydf7AF77XII7HZg:-739720581) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender_test.go#L340)):*
use util.Errorf for the line number

---

*[kv/txn_coord_sender_test.go, line 348 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfe4w60aOj9vPrdw4D:-KEfe4w60aOj9vPrdw4E:-285781465) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender_test.go#L348)):*
s/GoError/GetDetail/?

---

*[roachpb/data.go, line 634 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfVJ6yl4_aXyVAqv1e:-KEfVJ6yl4_aXyVAqv1f:528011846) ([raw file](https://github.com/cockroachdb/cockroach/blob/b95c879f17f6e8a9936d9692622d239797490410/roachpb/data.go#L634)):*
any reason to use `Forward` rather than just returning `*t.LastHeartbeat`? The former allocates.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tbg committed:
roachpb: remove BatchResponse.Header()

tschottdorf commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[kv/dist_sender.go, line 488 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfeOy_jLz0OCWFoTYT:-KEl65TB_NUILTURJHbH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1888be7ba43df1771ccd175da270518f3ccd2a2e/kv/dist_sender.go#L488)):*
Done.

---

*[kv/dist_sender.go, line 573 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEffMOzDaz8eLckb067:-KEl6OupOLAk9DuybR_o:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3220ac5b9a40c0ce6a3f9645c10978f055fe63d7/kv/dist_sender.go#L573)):*
Done.

---

*[kv/txn_coord_sender.go, line 643 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfd1U16RjKSeAFHBIS:-KEl7DJ4BwiigcEChHJ6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L643)):*
Done.

---

*[kv/txn_coord_sender.go, line 660 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdCSRfZ9gorckxyCE:-KEl7JCAEJpi9VTDQAth:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L660)):*
Done.

---

*[kv/txn_coord_sender.go, line 731 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdpUazETUX7lWtKjz:-KEl7WtxUinnEOYrFu1W:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L731)):*
Done.

---

*[kv/txn_coord_sender.go, line 742 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdjm7b4pwfxL9WuCN:-KEl8FcAW6PagG_Qvl3j:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender.go#L742)):*
Done.

---

*[kv/txn_coord_sender.go, line 736 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfelkfXZt_AwrBlwyR:-KEl7b-m-WkQLnReo68W:1166831932) ([raw file](https://github.com/cockroachdb/cockroach/blob/3220ac5b9a40c0ce6a3f9645c10978f055fe63d7/kv/txn_coord_sender.go#L736)):*
Debugging remnant. Removed.

---

*[kv/txn_coord_sender_test.go, line 340 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdydf7AF77XII7HZf:-KEl8PViTP3WXvqXeHhx:380713043) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender_test.go#L340)):*
SucceedsSoon already does it.

---

*[kv/txn_coord_sender_test.go, line 348 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfe4w60aOj9vPrdw4D:-KEl8Xc48p6UIeM89c54:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender_test.go#L348)):*
Done.

---

*[roachpb/data.go, line 634 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfVJ6yl4_aXyVAqv1e:-KEl8w7zujX8Hc-lftcp:-1240863436) ([raw file](https://github.com/cockroachdb/cockroach/blob/b95c879f17f6e8a9936d9692622d239797490410/roachpb/data.go#L634)):*
Changed, but does it actually allocate? @nvanbenschoten ;-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r6, 3 of 3 files at r7, 4 of 4 files at r8, 2 of 2 files at r9, 2 of 2 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[kv/txn_coord_sender_test.go, line 340 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdydf7AF77XII7HZf:-KElLxZ5vPsrSFGqaYD6:-985645466) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender_test.go#L340)):*
I don't think so - SucceedsSoon will give you its line number, not the line number of this `return`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[kv/txn_coord_sender_test.go, line 340 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdydf7AF77XII7HZf:-KEleGnOmVytv-sSlJtY:-551085854) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender_test.go#L340)):*
What it returns is more than enough to get to the heart of the problem - that's why it adds a line number. Having two almost adjacent line numbers shoved up next to each other is more confusing than helpful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[kv/txn_coord_sender_test.go, line 340 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEfdydf7AF77XII7HZf:-KElf1qBvqXSzPbh69uv:9328030) ([raw file](https://github.com/cockroachdb/cockroach/blob/5673d489b7ca0355a243760b8215765bc919cab6/kv/txn_coord_sender_test.go#L340)):*
Sure, it's enough, and yet having the precise line number is more convenient, especially when using click-to-open in iTerm.

Do what you feel, but this is a real annoyance for some (me).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@spencerkimball, ball in your court.

spencerkimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 334 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmfcxpuN-g8XbYW22A:-KEmfcxpuN-g8XbYW22B:1567961208) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L334)):*
Break this logic out into a new method; let's avoid massively long methods.

---

*[kv/txn_coord_sender.go, line 343 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmfF6c6ocdR4KYeCsC:-KEmfF6c6ocdR4KYeCsD:-1042778539) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L343)):*
Could we add an issue to add the coordinator node ID to the transaction session so that we can definitively return the right error between these two possible errors when we don't find the expected txn on the coordinator? Seems like a great help wanted feature.

---

*[kv/txn_coord_sender.go, line 689 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmguk8M2sE-TMBrA8g:-KEmguk8M2sE-TMBrA8h:-851027860) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L689)):*
s/not ongoing any more, we're just waiting/already finalized, so we wait/

---

*[kv/txn_coord_sender.go, line 730 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmhSckUcy0BSsdte9p:-KEmhSckUcy0BSsdte9q:447122004) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L730)):*
What's with the random TODO here?

---

*[kv/txn_coord_sender.go, line 736 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmhh10Y7Q-QxYEeIFy:-KEmhh118TAE-NjU1IiW:561547129) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L736)):*
s/stored proto/txn in the txns map/

s/give long-running transactions free updates/update long-running transactions/

---

*[roachpb/data.go, line 612 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmjcRTHzvDv6Bj0Z-v:-KEmjcRUR2IKhHGhmNuP:549060771) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/roachpb/data.go#L612)):*
s/occurred/definitively occurred/

---

*[storage/gc_queue.go, line 67 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmk1G9CJ6wJH4Cy5MW:-KEmk1G9CJ6wJH4Cy5MX:-353117277) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/gc_queue.go#L67)):*
Geez, 15 seems a bit arbitrarily high. 3x makes more sense, as long as we abort txns from the coordinator if they have not been successfully heartbeat within 2x the interval.

---

*[storage/replica_command.go, line 1011 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmnODiqhDQRf50IdxX:-KEmnODjjwynPJo3XXdR:136927926) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/replica_command.go#L1011)):*
s/allows to/allows it to/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-:-KEmnsrfGESRym_vlF7p:1417088721)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 6 of 7 files at r1, 1 of 3 files at r7, 1 of 4 files at r8, 2 of 2 files at r9, 2 of 2 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[kv/txn_coord_sender.go, line 349 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmnjQya2s1nBcWg7SZ:-KEmnjQya2s1nBcWg7S_:1853855297) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L349)):*
Is there any plausible way for the coordinator to learn that the transaction was committed via this path?

---

*[storage/replica_command.go, line 1159 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmls2z9coq4HkC0RrW:-KEmls2z9coq4HkC0RrX:391357532) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/replica_command.go#L1159)):*
This affects data written by a raft command, so it needs a migration plan.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tbg committed:
kv: move rejection code out of TxnCoordSender.Send

tschottdorf commented:
@bdarnell: I'm waiting for the merge because of the discussion on the migration. Let me know how you want to proceed.

---

Review status: 4 of 13 files reviewed at latest revision, 10 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 334 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmfcxpuN-g8XbYW22A:-KF8k7KJIlKUx1buejMr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L334)):*
Done.

---

*[kv/txn_coord_sender.go, line 343 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmfF6c6ocdR4KYeCsC:-KF8gxvuzxsa0XbPmmzL:1286673473) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L343)):*
I added a TODO in your name - seems like a minor point since we control the timeout via the context on the v3conn, and since changing coordinators will only happen if you're running a pgwire load-balancing solution which completely disrespects transaction blocks. I also added to the TODO another possibility, which is to just look at the difference between OrigTimestamp and the current coordinator clock which allows for an educated guess about a timeout having happened.

---

*[kv/txn_coord_sender.go, line 349 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmnjQya2s1nBcWg7SZ:-KF8hO65uQiZTVHRubU2:-528965740) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L349)):*
You could imagine that happening if `EndTransaction` failed due to network issues (but in fact did happen), and the client only retrying after another heartbeat happened (i.e. seconds later). The market for this is vanishingly small, so I didn't explore this further.

---

*[kv/txn_coord_sender.go, line 689 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmguk8M2sE-TMBrA8g:-KF8hneV8S8mFkOHvkAD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L689)):*
Done.

---

*[kv/txn_coord_sender.go, line 730 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmhSckUcy0BSsdte9p:-KF8i-PuYxnpzSOje3oB:-1023298865) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L730)):*
Remnant from chasing a data race. Removed.

---

*[kv/txn_coord_sender.go, line 736 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmhh10Y7Q-QxYEeIFy:-KF8iNKJ_8w9oHMqRwpH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/kv/txn_coord_sender.go#L736)):*
Done.

---

*[roachpb/data.go, line 612 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmjcRTHzvDv6Bj0Z-v:-KF8iUEG5OthHAUYHtli:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/roachpb/data.go#L612)):*
Done.

---

*[storage/gc_queue.go, line 67 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmk1G9CJ6wJH4Cy5MW:-KF8iWyIUAGQIUx7_9-h:-924054988) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/gc_queue.go#L67)):*
No need to be strict there - the interval is like what, 5 seconds? So this throws out things after 75 seconds. I changed it to `5*`; it's good to have a little bit of buffer with clock offsets and heartbeats getting delayed through network issues, and the like.

---

*[storage/replica_command.go, line 1011 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmnODiqhDQRf50IdxX:-KF8jH0qsccg54P8_ewR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/replica_command.go#L1011)):*
Done.

---

*[storage/replica_command.go, line 1159 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmls2z9coq4HkC0RrW:-KF8jYPdAVP5cFaAkJBV:-93958655) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/replica_command.go#L1159)):*
You're right. Can the migration plan be "it does not matter in practice unless the consistency queue complains about it"? The transaction is aborted, so its timestamp isn't relevant (except for this GC queue thing), and since the entry will be garbage-collected, the problem essentially solves itself (at least in the absence of a mixed-version cluster, which is not something we ought to be aiming at at the moment).
In the context of the open RFC #5985, this would be a harmless migration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-:-KF8lLwAMplG-gPhGFrg:341915935)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 8 files at r12, 2 of 9 files at r18, 1 of 3 files at r19, 1 of 4 files at r20, 1 of 2 files at r21, 1 of 1 files at r22, 2 of 2 files at r23, 1 of 1 files at r24.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 450 \[r24\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KFAiARpEhTVtZLGPeDq:-KFAiARpEhTVtZLGPeDr:-709028288) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd09544652976e37ef1136fa0defba5809c4e26/kv/txn_coord_sender.go#L450)):*
please follow the style guide

---

*[kv/txn_coord_sender.go, line 451 \[r24\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KFAituc_143M6QpgkH3:-KFAituc_143M6QpgkH4:524819522) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd09544652976e37ef1136fa0defba5809c4e26/kv/txn_coord_sender.go#L451)):*
why does this take a pointer?

---

*[kv/txn_coord_sender.go, line 765 \[r24\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KFAi4Gwn1rGN3uZZJoJ:-KFAi4GzRHedZzSMOOhF:161250711) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd09544652976e37ef1136fa0defba5809c4e26/kv/txn_coord_sender.go#L765)):*
random whitespace? suggest making this follow the style guide if you're going to touch it.

---

*[storage/replica_command.go, line 1011 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmnODiqhDQRf50IdxX:-KFAiQ9fkf-ZsB_gezKm:-1400386731) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/replica_command.go#L1011)):*
I think the old sentence was better; what is "it" now? This sentence's subject is "This" which is the action described in the previous sentence; that action does not make the decision described in this sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[storage/replica_command.go, line 1159 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmls2z9coq4HkC0RrW:-KFAi_6xLE4IeGhH2oPb:198037868) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/replica_command.go#L1159)):*
The problem solves itself if the GC queue runs before the consistency checker does, but there's still a window of inconsistency. If the consistency checker were running in panic mode then this would be a concern and we'd need to think more about the migration, but since we're not enabling consistency panics yet I think it's fine to let this merge as-is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 5 of 13 files reviewed at latest revision, 12 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 450 \[r24\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KFAiARpEhTVtZLGPeDq:-KFEHw-hHifN9mSKRQMH:-850546958) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd09544652976e37ef1136fa0defba5809c4e26/kv/txn_coord_sender.go#L450)):*
All stylish now.

---

*[kv/txn_coord_sender.go, line 451 \[r24\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KFAituc_143M6QpgkH3:-KFEK3mm61jyiXjkm_gm:950726544) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd09544652976e37ef1136fa0defba5809c4e26/kv/txn_coord_sender.go#L451)):*
Being sloppy. It now takes a `roachpb.Transaction`.

---

*[kv/txn_coord_sender.go, line 765 \[r24\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KFAi4Gwn1rGN3uZZJoJ:-KFEMEalf9BkezGYz886:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dd09544652976e37ef1136fa0defba5809c4e26/kv/txn_coord_sender.go#L765)):*
Done.

---

*[storage/replica_command.go, line 1011 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-KEmnODiqhDQRf50IdxX:-KFENW1S8CxesDpLw7xb:-470354389) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9a2bf87606d0889c661841bc4367864b28af265/storage/replica_command.go#L1011)):*
Reworded to something that's hopefully even better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 8 files at r25, 1 of 3 files at r26, 1 of 4 files at r27, 1 of 2 files at r28, 1 of 1 files at r29, 2 of 2 files at r30, 1 of 1 files at r31.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Any more comments here? Otherwise, merging in a few hours.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882#-:-KFFQJNOK0jWH0EPN2Kp:1126531641)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 9 files at r18, 1 of 8 files at r25, 1 of 3 files at r33, 1 of 4 files at r34, 1 of 2 files at r35, 1 of 1 files at r36, 2 of 2 files at r37, 1 of 1 files at r38.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
(I'll fix the data races before, of course).

tbg committed:
kv: remove potential for allocation

tschottdorf commented:
Ok, all fixed. Merging in +8h on green.

bdarnell commented:
Reviewed 1 of 8 files at r39, 1 of 3 files at r40, 1 of 4 files at r41, 1 of 2 files at r42, 1 of 1 files at r43, 2 of 2 files at r44, 1 of 1 files at r45.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5882)*
<!-- Sent from Reviewable.io -->
